### PR TITLE
make magic method new to follow cpython way

### PR DESCRIPTION
**Pull request by eldpswp99**
this issue was first mentioned in #2974 
This makes tp_new to be called by `object.__new__`

for example, with `tuple.__new__`

before: `<bound method __new__ of <class 'tuple'>>`
after: `<bound method object.__new__ of <class 'tuple'>>`

well, i have problem with controlling the new with pysuper, i remain this as draft now
I would be thankful for any advice!

### CODE REVIEW

File: vm/src/builtins/object.rs [hunk 1]
```diff
@@ -33,6 +34,11 @@ impl PyValue for PyBaseObject {
 
 #[pyimpl(flags(BASETYPE))]
 impl PyBaseObject {
+    #[pyclassmethod(magic)]
+    fn new(cls: PyTypeRef, args: FuncArgs, vm: &VirtualMachine) -> PyResult {
+        let (subtype, args): (PyTypeRef, FuncArgs) = args.bind(vm)?;
```
Comment (youknowone) [inline vm/src/builtins/object.rs:39]: `SlotConstructor` will be preferred for newly added `__new__` with binding.

Comment (eldpswp99) [reply vm/src/builtins/object.rs:39]: I thought there may be collision with object.tp_new. is it okay to put both?

Comment (youknowone) [reply vm/src/builtins/object.rs:39]: oh! I am sorry. you are right. this is not `tp_new` but just `__new__`.

File: vm/src/pyobject.rs [hunk 2]
```diff
@@ -144,7 +136,6 @@ impl PyContext {
             exceptions,
             int_cache_pool,
             string_cache,
-            tp_new_wrapper,
```
Comment (youknowone) [inline vm/src/pyobject.rs:147]: we need this field yet in this implementation

**Comment by youknowone** [comment]
`__new__` will not be looked up from tp_new anymore and the wrapper function `__new__` will be added to classes only when tp_new slot is defined.

**Comment by youknowone** [comment]
@coolreader18 @DimitrisJim could you review this PR?

File: vm/src/builtins/pytype.rs [hunk 3]
```diff
@@ -239,6 +239,19 @@ impl PyTypeRef {
 
 #[pyimpl(with(SlotGetattro, SlotSetattro, Callable), flags(BASETYPE))]
 impl PyType {
+    // bound method for every type
+    pub(crate) fn __new__(zelf: PyRef<Self>, args: FuncArgs, vm: &VirtualMachine) -> PyResult {
```
Comment (DimitrisJim) [inline vm/src/builtins/pytype.rs:243]: `__new__` is sort of confusing to me, I'd prefer if this kept `tp_new_wrapper` so we can easily see which function this maps to in CPython.

**Comment by DimitrisJim** [comment]
this looks just fine to me but I've not really looked into it much so I'll leave the approval for Noa.

File: vm/src/builtins/pytype.rs (hunk 3 above)
Comment (youknowone) [reply vm/src/builtins/pytype.rs:243]: This is a body of bound method `__new__` just like `mro`. The only difference of them is `__new__` is a bound method for every type including `type` but `mro` is an unbound method for `type` and bound as boundmethod for other types.
I think we'd better to keep this name as same as other methods when it works like other methods.

**Review by coolreader18 — APPROVED** [review]
LGTM, I think :sweat_smile: 

This whole part of the interpreter is pretty complex and I don't really have an intuition for how it works, so given that this seems simpler and still works it's got my approval :+1:

**Pull request merged**